_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+
+    @GetMapping("/")
+    public String index(@AuthenticationPrincipal OidcUser oidcUser, Model model) {
+        model.addAttribute("username", oidcUser.getEmail());
+        return "index";
+    }
+}
+```
+
+Add a `profile.html` template:
+
+```html
+<!-- src/main/resources/templates/profile.html -->
+<!doctype html>
+<html lang="en">
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Bootstrap demo</title>
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-T3c6CoIi6uLrA9TneNEoa7RxnatzjcDSCmG1MXxSR1GAsXEV/Dwwykc2MPK8M2HN" crossorigin="anonymous">
+    <link rel="stylesheet" href="styles.css">
+</head>
+<body>
+
+<th:block th:replace="~{fragments/navigation.html}" />
+
+
+<div class="container my-5">
+    <h1>[[${username}]]</h1>
+    <div class="col-lg-8 px-0">
+        <p class="fs-5">Here are your user's attributes:</p>
+        <table class="table table-striped">
+            <thead>
+            <tr>
+                <th>Claim</th>
+                <th>Value</th>
+            </tr>
+            </thead>
+            <tbody>
+            <tr th:each="item : ${claims}">
+                <td th:text="${item.key}">Key</td>
+                <td th:id="${'claim-' + item.key}" th:text="${item.value}">Value</td>
+            </tr>
+            </tbody>
+        </table>
+    </div>
+</div>
+
+<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-C6RzsynM9kWDrMNeT87bh95OGNyZPhcTNXj1NW7RuBCsyN/o0jlpcV8Qyq46cDfL" crossorigin="anonymous"></script>
+<script src="main.js"></script>
+</body>
+</html>
+```
+
+Add the `/profile` handler in the `HomeController` class:
+
+```java
+// src/main/java/com/example/demo/web/HomeController.java
+...
+
+@GetMapping("/profile")
+public String profile(@AuthenticationPrincipal OidcUser oidcUser, Model model) {
+    model.addAttribute("username", oidcUser.getEmail());
+    model.addAttribute("claims", oidcUser.getClaims());
+    return "profile";
+}
+
+...
+```
+
+### Configure Authentication Code Flow with PKCE at Auth0
+
+> Auth0 supports [OAuth 2.0 Security Best Current Practice](https://oauth.net/2/oauth-best-practice/), and you can register the application as a Regular Web Application (confidential client), and the Auth0 authorization server will honor the PKCE flow.
+
+Sign up at [Auth0](https://auth0.com/signup) and install the [Auth0 CLI](https://github.com/auth0/auth0-cli). Then in the command line run:
+
+```shell
+auth0 login
+```
+
+The command output will display a device confirmation code and open a browser session to activate the device.
+
+Using the Auth0 CLI, register the web application as an authentication client for Auht0:
+
+```shell
+auth0 apps create \
+  --name "Spring MVC" \
+  --description "Spring Boot Webapp" \
+  --type regular \
+  --callbacks http://localhost:4040/login/oauth2/code/okta \
```

Comment:
why 4040, isn't the default port for new SB apps 8080?

Reply:
Changed port.